### PR TITLE
adding sms endpoint

### PR DESCRIPTION
**Pull request by razanmedani**

### CODE REVIEW

File: consumer/auth.go [hunk 1]
```diff
@@ -193,7 +193,7 @@ func (s *State) CreateUser(c *gin.Context) {
 		panic("wtf")
 	}
 	if err := s.Db.AutoMigrate(&gateway.UserModel{}); err != nil {
-		log.Print("the error is: %v", err)
```
Comment (adonese) [inline consumer/auth.go:196]: We want to use log, not fmt. The logger include other valuable info such as date and time, unlike fmt

**Review by adonese — CHANGES_REQUESTED** [review]

File: consumer/helper.go [hunk 2]
```diff
@@ -197,7 +197,7 @@ func userExceededMaxSessions(s *State, username string) bool {
 			//s.Redis.HSet(username, "suspecious_behavior", 1)
 			s.Redis.HIncrBy(username, "suspicious_behavior", 1)
 			ttl, _ := s.Redis.TTL(username + ":login_counts").Result()
-			log.Printf("user exceeded max sessions", ttl)
```
Comment (adonese) [inline consumer/helper.go:200]: Ok this works

File: consumer/services.go [hunk 3]
```diff
@@ -19,6 +21,10 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
+const (
+	SMS_GATEWAY = "https://mazinhost.com/smsv1/sms/api?action=send-sms"
+)
+
```
Comment (adonese) [inline consumer/services.go:27]: Good use for static variables

File: consumer/services.go [hunk 4]
```diff
@@ -555,3 +561,53 @@ func (p *paymentTokens) pubSub(channel string, message interface{}) {
...
+//send sms to the user
+func (s *Service) SendSMS(c *gin.Context) {
+	// from gin, you read the request body (api documentations TODO)
+	var fields SMS
+	//var fields ebs_fields.MobileRedis
+	err := c.ShouldBindWith(&fields, binding.JSON)
+
+	if err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error(), "code": "unmarshalling_error"})
+	}
```
Comment (adonese) [inline consumer/services.go:574]: We could simplify this as if err := c.ShouldBindWith(..)

File: consumer/services.go [hunk 5]
```diff
@@ -555,3 +561,53 @@ func (p *paymentTokens) pubSub(channel string, message interface{}) {
...
+	var message string
+	phone := c.GetString("phone")
+	message = c.GetString("message")
+	if (phone == "") && (message == "") {
+		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error(), "code": "invalid values"})
+	}
+	fields.MobileNo = phone
+	fields.message = message
```
Comment (adonese) [inline consumer/services.go:582]: Nope, this doesn't work like this. You have to specify how your api endpoint would be used. `c.ShouldBindWith` binds the request body to a json object -- and only works for `POST` requests.

`c.GetString` get's a string from the context -- works when we want to by-pass data between middlewares, check `c.GetString("token")`.

Rewrite this code to something like this:

if err := c.ShouldBindWith // previous comment
// handle errors properly
if err := sendSms(fields) // we marshalled it earlier

Comment (adonese) [reply consumer/services.go:582]: this whole part is not needed

File: consumer/services.go [hunk 6]
```diff
@@ -555,3 +561,53 @@ func (p *paymentTokens) pubSub(channel string, message interface{}) {
...
+	v.Add(("action"), "send-sms")
+	v.Add("from", "Cashq") //change to tutipay?, do vendors need to change this from their side ?
+	v.Add("numbers", sms.MobileNo)
+	v.Add("message", sms.message) //contains the bill details (time/status/service-name/...)
+
+	url := SMS_GATEWAY + v.Encode()
+	res, err := http.Get(url)
+	if err != nil {
+		log.Printf("The error is: %v", err)
+
```
Comment (adonese) [inline consumer/services.go:609]: return the error here, or else the function would always return nil as error

File: consumer/services_test.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+	}{
+		// TODO: Add test cases.
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			tt.s.SendSMS(tt.args.c)
+		})
+	}
+}
+
```
Comment (adonese) [inline consumer/services_test.go:26]: Those are empty! write your actuall test in todo

File: consumer/services_test.go [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+		// TODO: Add test cases.
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if err := sendSMS(tt.args.sms); (err != nil) != tt.wantErr {
+				t.Errorf("sendSMS() error = %v, wantErr %v", err, tt.wantErr)
+			}
+		})
+	}
+}
```
Comment (adonese) [inline consumer/services_test.go:45]: Same here, i can populate the test files no problem

File: consumer/types.go [hunk 9]
```diff
@@ -774,3 +774,14 @@ func newFromBytes(d []byte, code int) (response, error) {
 		}, nil
 	}
 }
+
+type SMS struct {
+	MobileNo string `json:"mobileNo,omitempty"`
+	message  string `json:"message,omitempty"`
```
Comment (adonese) [inline consumer/types.go:780]: Fields that start with a lower case are not exported, they won't be marshalled

File: consumer/types.go [hunk 10]
```diff
@@ -774,3 +774,14 @@ func newFromBytes(d []byte, code int) (response, error) {
 		}, nil
 	}
 }
+
+type SMS struct {
+	MobileNo string `json:"mobileNo,omitempty"`
```
Comment (adonese) [inline consumer/types.go:779]: Change the field name to `Mobile` and the marshall to just `mobile`. We try to keep fields as small as possible, and we use `type_face` for json objects (except for ebs compatibility)

File: consumer/types.go [hunk 11]
```diff
@@ -774,3 +774,14 @@ func newFromBytes(d []byte, code int) (response, error) {
...
+	message  string `json:"message,omitempty"`
+
+	//	TranDateTime string `json:"tranDateTime,omitempty"` // i will embed all these dtails in the front-end
+	//  TranAmount   string `json:"tranAmount,omitempty"`
+	//	PAN          string `json:"PAN,omitempty"`
+	//	FromAccount  string `json:"fromAccount,omitempty"`
+	//	ToAccount    string `json:"toAccount,omitempty"`
```
Comment (adonese) [inline consumer/types.go:786]: Delete comment files

File: ebs_service.go [hunk 12]
```diff
@@ -179,6 +179,7 @@ func GetMainEngine() *gin.Engine {
 	mGroup.POST("/login", merchantServices.Login)
 	mGroup.POST("/m", merchantServices.AddBilling)
 	mGroup.PUT("/update", merchantServices.Update)
+	mGroup.POST("/sms", consumerService.SendSMS)
```
Comment (adonese) [inline ebs_service.go:182]: Write documentations for this new api

**Review by adonese — CHANGES_REQUESTED** [review]

File: consumer/auth.go (hunk 1 above)
Comment (razanmedani) [reply consumer/auth.go:196]: okaay .. i just got an error when i performed the testing that says (Print call has possible formatting directive %v) and when i searched about it i figured that Print does not support formatting directives.

Comment (adonese) [reply consumer/auth.go:196]: Nope that is for %# you added in a latter place

File: consumer/services.go [hunk 13]
```diff
@@ -555,3 +561,47 @@ func (p *paymentTokens) pubSub(channel string, message interface{}) {
...
+	 https://mazinhost.com/smsv1/sms/api?action=send-sms&
+	 api_key=bW1idXNpZkBnbWFpbC5jb206bURTMVJQc2d1JA==
+	 &to=249111493885
+	 &from=COOP
+	 &sms="you here"
+	*/
+	v := url.Values{}
+	v.Add("api_key", os.Getenv("SMS_API_KEY")) // supply it in ENV
+	v.Add(("action"), "send-sms")
+	v.Add("from", "Cashq") //change to tutipay?, do vendors need to change this from their side ?
```
Comment (adonese) [inline consumer/services.go:594]: Good point, we can configure it later.

**Review by adonese — COMMENTED** [review]
This is really neat!

**Review by adonese — APPROVED** [review]

**Review by adonese — APPROVED** [review]

**Pull request merged**